File: lib/promptlib/rendering/template_renderer.py
"""`{{variable}}` substitution over a prompt body.

Chosen over str.format because prompt bodies routinely contain braces (JSON
examples, code) that format() would misparse. The double-brace form is rare in
prose and matches the convention already used across this user's prompt corpus.
"""

from __future__ import annotations

import re
from dataclasses import dataclass
from typing import Mapping

from ..errors import RenderError

_PLACEHOLDER = re.compile(r"\{\{\s*([A-Za-z0-9_.-]+)\s*\}\}")


@dataclass(frozen=True)
class RenderResult:
    """Outcome of a substitution pass."""

    text: str
    substituted: tuple[str, ...]
    unfilled: tuple[str, ...]

    def to_dict(self) -> dict[str, object]:
        return {
            "text": self.text,
            "substituted": list(self.substituted),
            "unfilled": list(self.unfilled),
        }


class TemplateRenderer:
    """Substitutes `{{name}}` placeholders from a value mapping."""

    def placeholders(self, template: str) -> tuple[str, ...]:
        """Return placeholder names in first-appearance order."""
        seen: list[str] = []
        for match in _PLACEHOLDER.finditer(template):
            name = match.group(1)
            if name not in seen:
                seen.append(name)
        return tuple(seen)

    def render(
        self,
        template: str,
        values: Mapping[str, object],
        strict: bool = False,
    ) -> RenderResult:
        """Fill placeholders; in strict mode a missing value is an error."""
        substituted: list[str] = []
        unfilled: list[str] = []

        def replace(match: re.Match[str]) -> str:
            name = match.group(1)
            if name in values and values[name] is not None:
                if name not in substituted:
                    substituted.append(name)
                return str(values[name])
            if name not in unfilled:
                unfilled.append(name)
            return match.group(0)

        text = _PLACEHOLDER.sub(replace, template)
        if strict and unfilled:
            raise RenderError(f"missing values for: {', '.join(unfilled)}")
        return RenderResult(text=text, substituted=tuple(substituted), unfilled=tuple(unfilled))

File: site/src/application/template-renderer.ts
import type { PromptPart } from '../events/prompt-part.ts';

/** `{{variable}}` substitution, as lib/promptlib/rendering/template_renderer.py does it. */
const PLACEHOLDER = /\{\{\s*([A-Za-z0-9_.-]+)\s*\}\}/g;

export class TemplateRenderer {
  static parts(template: string): PromptPart[] {
    const parts: PromptPart[] = [];
    let last = 0;
    for (const match of template.matchAll(PLACEHOLDER)) {
      if (match.index > last) parts.push({ kind: 'text', text: template.slice(last, match.index) });
      parts.push({ kind: 'variable', text: match[1] });
      last = match.index + match[0].length;
    }
    if (last < template.length) parts.push({ kind: 'text', text: template.slice(last) });
    return parts;
  }

  static render(template: string, values: Readonly<Record<string, string>>): { text: string; unfilled: string[] } {
    const unfilled: string[] = [];
    const text = template.replace(PLACEHOLDER, (whole, name: string) => {
      const value = values[name];
      if (value) return value;
      if (!unfilled.includes(name)) unfilled.push(name);
      return whole;
    });
    return { text, unfilled };
  }
}
